src/ucp/core/ucp_mm.c

```diff
@@ -291,100 +273,200 @@ static inline int ucp_mem_map_is_allocate(const ucp_mem_map_params_t *params)
-    if (memh->alloc_method != UCT_ALLOC_METHOD_LAST) {
-        mem.address = memh->address;
-        mem.length  = memh->length;
-        mem.method  = memh->alloc_method;
-        mem.md      = memh->alloc_md;  /* May be NULL if method is not MD */
-        mem.memh    = alloc_md_memh;   /* May be INVALID if method is not MD */
+    if (alloc) {
+        if (mem.method == UCT_ALLOC_METHOD_MD) {
+            for (md_index = 0; md_index < context->num_mds; md_index++) {
+                if (mem.md == context->tl_mds[md_index].md) {
+                    region->memh.alloc_md_index = md_index;
+                    region->memh.uct[md_index]  = mem.memh;
+                    region->memh.md_map        |= UCS_BIT(md_index);
+                    break;
+                }
+            }
+        }
 
-        status = uct_mem_free(&mem);
-        if (status != UCS_OK) {
-            goto out;
+        region->memh.alloc_method = mem.method;
+        region->memh.mem_type     = mem.mem_type;
+    } else {
+        region->memh.mem_type     = mem_type;
+    }
+
+    new_md_map = ~region->memh.md_map & reg_md_map;
+    new_md_map_ok = 0;
+    ucs_for_each_bit(md_index, new_md_map) {
+        status = uct_md_mem_reg(context->tl_mds[md_index].md,
+                                reg_address, reg_length, uct_flags,
+                                &region->memh.uct[md_index]);
+        if (ucs_unlikely(status != UCS_OK)) {
+            log_level = (uct_flags & UCT_MD_MEM_FLAG_HIDE_ERRORS) ?
+                                UCS_LOG_LEVEL_DIAG :
+                                UCS_LOG_LEVEL_ERROR;
+            ucs_log(log_level,
+                    "failed to register %s %p length %zu on md[%d]=%s: %s",
+                    ucs_memory_type_names[mem_type], reg_address, reg_length,
+                    md_index, context->tl_mds[md_index].rsc.md_name,
+                    ucs_status_string(status));
+            ucp_memh_release(context, region, new_md_map_ok);
+
+            if (context->rcache != NULL) {
+                ucs_rcache_region_put(context->rcache, &region->super);
+            } else {
+                ucs_free(region);
+            }
+
+            return status;
         }
+
+        ucs_trace("registered address %p length %zu on md[%d]=%s %p",
+                  reg_address, reg_length, md_index,
+                  context->tl_mds[md_index].rsc.md_name,
+                  region->memh.uct[md_index]);
+        new_md_map_ok |= UCS_BIT(md_index);
     }
 
-    ucs_free(memh);
-    status = UCS_OK;
-out:
-    return status;
+    region->memh.md_map |= new_md_map_ok;
+    *memh_p = &region->memh;
+    return UCS_OK;
+}
+
+static void ucp_memh_put_common(ucp_context_h context, ucp_mem_h memh, int inv)
+{
+    ucp_mem_rcache_region_t *region = ucs_container_of(memh,
+                                                       ucp_mem_rcache_region_t,
+                                                       memh);
+    if (ucp_memh_zero_length(memh)) {
```

Comment:
1. ucp_memh_is_zero_length
2. unlikely
3. any way we can avoid this branch? maybe inf refcount on the dummy.super region?